src/couch_mrview/src/couch_mrview_http.erl

```diff
@@ -709,3 +750,169 @@ maybe_set_page_size(#mrargs{page_size = undefined} = Args, MaxPageSize) ->
+    Binary = couch_util:decodeBase64Url(Bookmark),
+    Term = binary_to_term(Binary, [safe]),
+    {Args, _} = lists:foldl(fun(Value, {Acc, Idx}) ->
+        case {Value, element(Idx, Acc)} of
+            {[], []} ->
+                {Acc, Idx + 1};
+            {[], Default} ->
+                {setelement(Idx, Acc, Default), Idx + 1};
+            {_, _} when Idx == #mrargs.bookmark ->
+                {setelement(Idx, Acc, true), Idx + 1};
+            {Value, _} ->
+                {setelement(Idx, Acc, Value), Idx + 1}
+        end
+    end, {#mrargs{}, 1}, [mrargs | tuple_to_list(Term)]),
+    Args.
+
+
+paginated(Req, EtagTerm, #mrargs{page_size = PageSize} = Args, KeyFun, Fun) ->
+    Etag = couch_httpd:make_etag(EtagTerm),
+    chttpd:etag_respond(Req, Etag, fun() ->
+        hd(paginated(PageSize, [set_limit(Args)], KeyFun, Fun))
+    end).
+
+
+paginated(Req, EtagTerm, PageSize, QueriesArgs, KeyFun, Fun) when is_list(QueriesArgs) ->
+    Etag = couch_httpd:make_etag(EtagTerm),
+    chttpd:etag_respond(Req, Etag, fun() ->
+        Results = paginated(PageSize, QueriesArgs, KeyFun, Fun),
+        #{results => Results}
+    end).
+
+paginated(PageSize, QueriesArgs, KeyFun, Fun) when is_list(QueriesArgs) ->
+    {_N, Results} = lists:foldr(fun(Args0, {Limit, Acc}) ->
+        case Limit > 0 of
+            true ->
+                Args = set_limit(Args0#mrargs{page_size = Limit}),
+                {Meta, Items} = Fun(Args),
+                Result = maybe_add_bookmark(
+                    PageSize, Args, Meta, Items, KeyFun),
+                #{total_rows := Total} = Result,
+                {Limit - Total, [Result | Acc]};
+            false ->
+                Bookmark = bookmark_encode(Args0),
+                Result = #{
+                    rows => [],
+                    next => Bookmark,
+                    total_rows => 0
+                },
+                {Limit, [Result | Acc]}
+        end
+    end, {PageSize, []}, QueriesArgs),
+    Results.
+
+
+maybe_add_bookmark(PageSize, Args0, Response, Items, KeyFun) ->
+    #mrargs{page_size = Limit} = Args0,
+    Args = Args0#mrargs{page_size = PageSize},
+    case check_completion(Limit, Items) of
+        {Rows, nil} ->
+            maps:merge(Response, #{
+                rows => Rows,
+                total_rows => length(Rows)
+             });
+        {Rows, Next} ->
+            NextKey = KeyFun(Next),
+            is_binary(NextKey)
+                orelse throw("Provided KeyFun should return binary"),
+            Bookmark = case Args#mrargs.keys of
+                undefined ->
+                    bookmark_encode(Args#mrargs{start_key=NextKey});
+                Keys when is_list(Keys) ->
+                    NewKeys = lists:foldl(fun(Row, Acc) ->
+                        gb_sets:del_element(KeyFun(Row), Acc)
+                    end, gb_sets:from_list(Keys), Rows),
+                    bookmark_encode(Args#mrargs{keys=gb_sets:to_list(NewKeys)})
```

Comment:
A couple questions here. First, why are we bookmarking pages of POST'ed keys? Seems like that's more of a 400 error sort of situation. Paging through a list of keys that the client has in RAM seems like a bit of a waste. I guess I could be convinced if it makes client libraries easier to implement but I'm really not seeing how that'd be the case off the top of my head.

Second, the gb_sets approach seems not right here. First, I'm not sure I get why you've gone with gb_sets here over sets or ordsets even. At first I thought it might retain ordering of the input list but that appears to not be the case:

```erlang
2> gb_sets:to_list(gb_sets:from_list([a, b, c, d])).
[a,b,c,d]
3> gb_sets:to_list(gb_sets:from_list([a, c, d, b])).
[a,b,c,d]
```

Secondly, this seems to be also changing the API contract as you're deleting keys from a set. My memory is a bit hazy, but I'm pretty sure if a user provides a docid/key twice then we return that row twice. And we retain their ordering (subject to query string parameters like descending/skip/limit).

Reply:
Would you propose to include keys into a bookmark as is and copy to bookmark for each next page?

Reply:
I was more thinking that we'd reject a paginated request with keys. But other than that I'd just drop them and include directly. Otherwise we've got two different result sets depending on paginated vs not.